zkvm/src/types.rs

```diff
@@ -320,10 +327,10 @@ impl Data {
         match self {
             Data::Opaque(data) => {
                 let point = Subslice::new(&data).read_point()?;
-                Ok(Predicate::Opaque(point))
+                Ok(Predicate::new(point))
             }
             Data::Witness(witness) => match witness {
-                DataWitness::Predicate(w) => Ok(Predicate::Witness(w)),
+                DataWitness::Predicate(w) => Ok(Predicate::new_witness(w.to_point(), *w)),
```

Comment:
This witness-to-point conversion should be failable, and its computation from witness is better inside Predicate::from_witness method (failable) that ensures that the point is in sync with the witness. I’d also keep Predicate.to_point method (non failable) to make witness and point private fields. We should probably use pub(crate) very rarely only when we can make it pub, but not sure if we should, while keeping it public within crate only.